pkg/controller/deployment/deployment_controller.go

```diff
@@ -557,9 +597,23 @@ func (dc *DeploymentController) getNewRC(deployment extensions.Deployment) (*api
 		dc.rcExpectations.DeleteExpectations(dKey)
 		return nil, fmt.Errorf("error creating replication controller: %v", err)
 	}
+
+	if err = dc.updateDeploymentVersion(deployment, newVersion); err != nil {
+		return createdRC, err
+	}
+
 	return createdRC, nil
 }
 
+func (dc *DeploymentController) updateDeploymentVersion(deployment extensions.Deployment, version string) error {
+	if deployment.Annotations == nil {
+		deployment.Annotations = make(map[string]string)
+	}
+	deployment.Annotations[deploymentutil.VersionAnnotation] = version
```

Comment:
> generation would reset to 0 after deleting and re-creating the Deployment

The deployment controller could observe any adoptions and set the correct generation number back in the deployment.